pkg/server/server.go

```diff
@@ -396,14 +445,32 @@ func (s *ProxyServer) Proxy(stream client.ProxyService_ProxyServer) error {
 		return fmt.Errorf("failed to get context")
 	}
 	userAgent := md.Get(header.UserAgent)
-	klog.V(5).InfoS("Proxy request from client", "userAgent", userAgent, "serverID", s.serverID)
+	streamUID := uuid.New().String()
+	klog.V(5).InfoS("Proxy request from client", "userAgent", userAgent, "serverID", s.serverID, "streamUID", streamUID)
 
 	recvCh := make(chan *client.Packet, xfrChannelSize)
 	stopCh := make(chan error, 1)
 
 	frontend := GrpcFrontend{
-		stream: stream,
+		stream:    stream,
+		streamUID: streamUID,
 	}
+
+	defer func() {
+		// The frontend stream and goroutines are completely shut down, so we will never get a
+		// subsequent close message from the frontend. This can happen for several reasons, such as
+		// stream canceled. Clean any lingering tunnel connections to avoid resource leaks.
+		for _, p := range s.PendingDial.RemoveForStream(streamUID) {
+			klog.V(2).InfoS("frontend stream shutdown, cleaning dial", "dialID", p.dialID)
+			// TODO: add agent support to handle this
+			s.sendBackendDialClose(p.backend, p.dialID, "frontend stream shutdown")
+		}
+		for _, f := range s.removeFrontendsForStream(streamUID) {
+			klog.V(2).InfoS("frontend stream shutdown, cleaning frontend", "connectionID", f.connectID, "dialID", f.dialID)
+			s.sendBackendClose(f.backend, f.connectID, f.dialID, "frontend stream shutdown")
```

Comment:
It probably doesn't matter much while there's a single connection, but should this (and `sendBackendDialClose` above) be called in a go routine, so they don't block terminating the frontend connection?

Reply:
I guess that would undo the change that prevents this method from returning until all associated go-routines have terminated...

Reply:
I could imagine continuing to block Proxy returning on both stream handling goroutines shutting down, and "leak" one best effort cleanup goroutine.

But I prefer not, because it makes unit tests a bit more non-deterministic. And there is something nice about "all work is done" on Proxy finish (and monitoring the Proxy gauge metric).

Reply:
The only issue would be if the backend stream was backed up, then send could block, and hold this connection open. I doubt that would turn into a permanent leak though...